File: legacy_restore/src_local/utils/maintenance_engine_phd.py
import logging
import json
import subprocess
import os

logger = logging.getLogger(__name__)

class MaintenanceEnginePhd:
    """🛠️ Motor de Manutenção PhD: Limpeza e Fusão de Índices."""
    
    @staticmethod
    def clean_submodules(root, run_git):
        if not (root / ".gitmodules").exists(): return
        res = run_git(["submodule", "foreach", "--quiet", "echo $displaypath"])
        if res and res.returncode == 0:
            for sub in res.stdout.splitlines():
                sub_dir = root / sub.strip()
                if sub_dir.exists(): subprocess.run(["git", "clean", "-fd"], cwd=str(sub_dir), capture_output=True)

    @staticmethod
    def merge_skills_index(root, file_path, run_git, protected_ids):
        ours_raw = run_git(["show", f":2:{file_path}"]).stdout
        theirs_raw = run_git(["show", f":3:{file_path}"]).stdout
        try:
            ours = json.loads(ours_raw) if ours_raw else []
            theirs = json.loads(theirs_raw) if theirs_raw else []
            merged = {item['id']: item for item in theirs}
            for item in ours:
                if item['id'] in protected_ids: merged[item['id']] = item
            with open(root / file_path, "w", encoding="utf-8") as f:
                json.dump(sorted(merged.values(), key=lambda x: x['id']), f, indent=2, ensure_ascii=False)
            return True
        except: return False

File: src_local/utils/maintenance_engine_phd.ts

import { join } from "path";
import { existsSync } from "fs";
import winston from "winston";
import { GitClient } from "./git_client";

const logger = winston.child({ module: "MaintenanceEnginePhD" });

export class MaintenanceEnginePhd {
    /**
     * Limpeza recursiva de submódulos.
     */
    static async cleanSubmodules(root: string, git: GitClient): Promise<void> {
        const displayPaths = await git.run(["submodule", "foreach", "--quiet", "echo $displaypath"]);

        if (displayPaths && displayPaths.exitCode === 0 && displayPaths.stdout) {
            const submodules = displayPaths.stdout.split("\n").filter(s => s.trim() !== "");
            for (const sub of submodules) {
                await this.cleanSingleSubmodule(root, sub.trim());
            }
        }
    }

    private static async cleanSingleSubmodule(root: string, sub: string): Promise<void> {
        const subPath = join(root, sub);
        if (existsSync(subPath)) {
            logger.info(`🧹 [Maintenance] Limpando submódulo: ${sub}`);
            await Bun.spawn(["git", "clean", "-fd"], { cwd: subPath }).exited;
        }
    }

    /**
     * Merge inteligente do índice de skills (JSON), preservando IDs protegidos.
     */
    static async mergeSkillsIndex(root: string, filePath: string, git: GitClient, protectedIds: string[]): Promise<boolean> {
        try {
            const ours = await this.getGitVersion(root, filePath, 2);
            const theirs = await this.getGitVersion(root, filePath, 3);

            const mergedMap = this.performSkillsMerge(ours, theirs, protectedIds);
            const result = Array.from(mergedMap.values()).sort((a, b) => (a.id || "").localeCompare(b.id || ""));

            await Bun.write(join(root, filePath), JSON.stringify(result, null, 2));
            return true;
        } catch (error) {
            logger.error(`❌ [Maintenance] Falha no merge de skills: ${error}`);
            return false;
        }
    }

    private static async getGitVersion(root: string, filePath: string, stage: number): Promise<any[]> {
        try {
            const res = await Bun.spawn(["git", "show", `:${stage}:${filePath}`], { cwd: root, stdout: "pipe" });
            const text = await new Response(res.stdout).text();
            return JSON.parse(text);
        } catch {
            logger.debug(`⚠️ Maintenance: Failed to parse stage ${stage} version of ${filePath}.`);
            return [];
        }
    }

    private static performSkillsMerge(ours: any[], theirs: any[], protectedIds: string[]): Map<string, any> {
        const mergedMap = new Map<string, any>();

        for (const item of theirs) {
            this.setIfHasId(mergedMap, item);
        }

        for (const item of ours) {
            this.setIfProtected(mergedMap, item, protectedIds);
        }

        return mergedMap;
    }

    private static setIfHasId(map: Map<string, any>, item: any) {
        if (item.id) map.set(item.id, item);
    }

    private static setIfProtected(map: Map<string, any>, item: any, protectedIds: string[]) {
        if (item.id && protectedIds.includes(item.id)) {
            map.set(item.id, item);
        }
    }
}
